source/agora/node/LederOutputStream.d

```diff
@@ -0,0 +1,130 @@
+/*******************************************************************************
+
+    Function definition and helper related to serialization
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.node.LedgerOutputStream;
+
+import agora.common.Transaction;
+
+/// Ditto
+public class LedgerOutputStream
+{
+    private ubyte[] data;
+
+    private size_t offset;
+
+    /// Ctor
+    public this () @trusted
+    {
+        data.length = 0;
+        offset = 0;
+    }
+
+    // Offset must be less than or equal to the length of the data.
+    invariant()
+    {
+        assert(offset <= data.length);
+    }
+
+    // Returns the data from start to offset.
+    ubyte[] toBytes() @trusted
+    {
+        return data[0 .. offset];
+    }
+
+    // Allocate the data size by nbytes.
+    void reserve(size_t nbytes) @trusted
+    in
+    {
+        assert(offset + nbytes >= offset);
+    }
+    out
+    {
+        assert(offset + nbytes <= data.length);
+    }
+    body
+    {
+        if(data.length < offset + nbytes)
+        {
+            void[] vdata = data;
+            vdata.length = (offset + nbytes);
+            data = cast(ubyte[]) vdata;
+        }
+    }
+
+    // Assign additional data size bytes length and store bytes in data
+    // and add offset location bytes size.
+    void write(in ubyte[] bytes) @trusted
+    {
+        reserve(bytes.length);
+        data[offset .. offset + bytes.length] = bytes[];
+        offset += bytes.length;
+    }
+
+    void write(in byte[] bytes) @trusted
+    {
+        write(cast(ubyte[]) bytes);
+    }
+
+    void writeLong(const long value) @trusted
+    {
+        reserve(long.sizeof);
+        *cast(long*)&data[offset] = value;
+        offset += long.sizeof;
+    }
+
+    void writeAmount(const Amount value) @trusted
+    {
+        writeLong(cast(long)value);
+    }
+
+    void writeUint(const uint value) @trusted
+    {
+        reserve(uint.sizeof);
+        *cast(uint*)&data[offset] = value;
+        offset += uint.sizeof;
+    }
+
+    void writeUlong(const ulong value) @trusted
+    {
+        reserve(ulong.sizeof);
+        *cast(ulong*)&data[offset] = value;
+        offset += ulong.sizeof;
+    }
+
+    // get Data bytes
+    public ubyte[] getData() @trusted
+    {
+        return this.data;
+    }
+
+    // TOBE
+    public void flush() @trusted
+    {
+        data.length = 0;
+        offset = 0;
+    }
+
+    // TOBE
+    public void clear() @trusted
+    {
+        data.length = 0;
+        offset = 0;
+    }
+
+    // TOBE
+    public void destroy() @trusted
```

Comment:
Having 3 functions with different name that do the same thing is a red flag ;)

Reply:
The function is in development, so this PR has been deleted.